File: android/apps/AutoTesterAndroid/app/src/main/java/com/mousebirdconsulting/autotester/TestCases/CustomBNGTileSource.java
package com.mousebirdconsulting.autotester.TestCases;

import android.app.Activity;

import com.mousebird.maply.BaseController;
import com.mousebird.maply.CoordSystem;
import com.mousebird.maply.GlobeController;
import com.mousebird.maply.ImageLoaderInterpreter;
import com.mousebird.maply.MapController;
import com.mousebird.maply.Mbr;
import com.mousebird.maply.OvlDebugImageLoaderInterpreter;
import com.mousebird.maply.Point2d;
import com.mousebird.maply.Proj4CoordSystem;
import com.mousebird.maply.QuadImageLoader;
import com.mousebird.maply.SamplingParams;
import com.mousebirdconsulting.autotester.Framework.MaplyTestCase;
import java.io.File;

/**
 * Created by sjg on 2/13/16.
 */
public class CustomBNGTileSource extends MaplyTestCase
{
    public CustomBNGTileSource(Activity activity) {
        super(activity, "British National Grid");
        this.setDelay(2);
    }

    // Put together a British National Grid system
    static public CoordSystem MakeBNGCoordSystem(Activity activity, boolean displayVersion)
    {
        File bngFile = copyAssetFile(activity, "bng/OSTN02_NTv2.gsb", "bng", "OSTN02_NTv2.gsb");
        String gridRef = " +nadgrids=" + bngFile;

        String projStr = "+proj=tmerc +lat_0=49 +lon_0=-2 +k=0.9996012717 +x_0=400000 +y_0=-100000 +ellps=airy +units=m +no_defs" + gridRef;
        Proj4CoordSystem coordSys = new Proj4CoordSystem(projStr);

        // Set the bounding box for validity.  It assumes it can go everywhere by default
        Mbr bbox = new Mbr();
        bbox.addPoint(new Point2d(1393.0196,13494.9764));
        bbox.addPoint(new Point2d(671196.3657,1230275.0454));

        // Now expand it out so we can see the whole of the UK
        if (displayVersion)
        {
            // Note: There may be a center/offset problem with the bounds.  Made them bigger to compensate
            double spanX = bbox.ur.getX() - bbox.ll.getX();
            double spanY = bbox.ur.getY() - bbox.ll.getY();
            double extra = 1.5;
            bbox.addPoint(new Point2d(-extra*spanX,-extra*spanY));
            bbox.addPoint(new Point2d(extra*spanX,extra*spanY));
        }

        coordSys.setBounds(bbox);

        return coordSys;
    }

    public static QuadImageLoader makeTestLoader(Activity activity, BaseController viewC, boolean displayVersion)
    {
        CoordSystem bngCoordSystem = MakeBNGCoordSystem(activity, displayVersion);

        SamplingParams params = new SamplingParams();
        params.setCoordSystem(bngCoordSystem);
        params.setCoverPoles(false);
        params.setEdgeMatching(false);
        params.setMinZoom(0);
        params.setMaxZoom(10);
        //params.setMinImportance(256 * 256);

        QuadImageLoader loader = new QuadImageLoader(params,null,viewC);
        loader.setBaseDrawPriority(1000);
        //loader.setLoaderInterpreter(new ImageLoaderInterpreter());

        OvlDebugImageLoaderInterpreter interp = new OvlDebugImageLoaderInterpreter();
        interp.setParentInterpreter(new ImageLoaderInterpreter());
        loader.setLoaderInterpreter(interp);

        //baseLayer.setImportanceScale(4.f);
        //Mbr bounds = bngCoordSystem.getBounds();
        //viewC.setViewExtents(bounds.ll, bounds.ur);

        return loader;
    }

    @Override
    public boolean setUpWithGlobe(GlobeController globeVC) throws Exception
    {
        baseCase = new StamenRemoteTestCase(getActivity());
        baseCase.setUpWithGlobe(globeVC);

        loader = makeTestLoader(getActivity(), globeVC,true);

        Point2d pt = Point2d.FromDegrees(-0.1275, 51.507222);
        globeVC.setPositionGeo(pt.getX(), pt.getY(), 0.4);

        return true;
    }

    @Override
    public boolean setUpWithMap(MapController mapVC) throws Exception
    {
        baseCase = new StamenRemoteTestCase(getActivity());
        baseCase.setUpWithMap(mapVC);

        loader = makeTestLoader(getActivity(), mapVC,true);

        Point2d pt = Point2d.FromDegrees(-0.1275, 51.507222);
        mapVC.setPositionGeo(pt.getX(), pt.getY(), 0.4);

        return true;
    }

    @Override
    public void shutdown() {
        if (loader != null) {
            loader.shutdown();
            loader = null;
        }
        if (baseCase != null) {
            baseCase.shutdown();
        }
        super.shutdown();
    }

    private MaplyTestCase baseCase = null;
    private QuadImageLoader loader = null;
}

File: android/apps/AutoTesterAndroid/app/src/main/java/com/mousebirdconsulting/autotester/TestCases/CustomBNGTileSource.kt
/*  CustomBNGTileSource.kt
 *  WhirlyGlobeLib
 *
 *  Created by Steve Gifford on 2/13/16.
 *  Copyright 2016-2021 mousebird consulting
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.mousebirdconsulting.autotester.TestCases

import android.app.Activity
import com.mousebird.maply.*
import com.mousebird.maply.TestTileFetcher.TestTileInfo
import com.mousebirdconsulting.autotester.Framework.MaplyTestCase
import kotlin.math.log10

open class CustomBNGTileSource : MaplyTestCase {
    constructor(activity: Activity?) :
            super(activity, "British National Grid Tile Source") {
        setDelay(2)
    }
    
    protected constructor(activity: Activity?, testName: String?, impl: TestExecutionImplementation?) :
            super(activity, testName, impl)
    
    @Throws(Exception::class)
    override fun setUpWithGlobe(globeVC: GlobeController): Boolean {
        baseCase = GeographyClass(activity).also {
            it.setUpWithGlobe(globeVC)
        }

        loader = makeTestLoader(activity, globeVC)

        val bound = geoBound(makeBNGCoordSystem(activity, false))
        val middle = bound.middle()
        val h = globeVC.findHeightToViewBounds(bound, middle)
        globeVC.setPositionGeo(middle, h/2)
        globeVC.animatePositionGeo(middle, h*1.1, 0.0, 1.0)

        return true
    }
    
    @Throws(Exception::class)
    override fun setUpWithMap(mapVC: MapController): Boolean {
        baseCase = GeographyClass(activity).also {
            it.setUpWithMap(mapVC)
        }

        loader = makeTestLoader(activity, mapVC)

        val bound = geoBound(makeBNGCoordSystem(activity, false))
        val middle = bound.middle()
        val h = mapVC.findHeightToViewBounds(bound, middle)
        mapVC.setPositionGeo(middle, h/3)
        mapVC.animatePositionGeo(middle, h, 0.0, 1.0)

        return true
    }
    
    private fun geoBound(coordSys: CoordSystem): Mbr {
        val bound = coordSys.bounds
        val ll = coordSys.localToGeographic(Point3d(bound.ll.x, bound.ll.y, 0.0))
        val ur = coordSys.localToGeographic(Point3d(bound.ur.x, bound.ur.y, 0.0))
        return Mbr(Point2d(ll.x, ll.y), Point2d(ur.x, ur.y))
    }
    
    override fun shutdown() {
        loader?.let {
            it.tileFetcher?.shutdown()
            it.shutdown()
            loader = null
        }
        baseCase?.let {
            it.shutdown()
            baseCase = null
        }
        super.shutdown()
    }

    // Put together a British National Grid system
    protected fun makeBNGCoordSystem(activity: Activity?, displayVersion: Boolean): CoordSystem {
        // Load up the grid adjustment file from the assets, and write it out where proj.4 can read it
        // todo: how do we quote/escape this if there are spaces?
        val bngFile = copyAssetFile(activity, "bng/OSTN02_NTv2.gsb", "bng", "OSTN02_NTv2.gsb")
        val projStr = "+proj=tmerc +lat_0=49 +lon_0=-2 +k=0.9996012717 +x_0=400000 +y_0=-100000 +ellps=airy +units=m +no_defs +nadgrids=$bngFile"
        
        // Set the bounding box for validity.  It assumes it can go everywhere by default
        val bound = Mbr().also {
            it.addPoint(Point2d(1393.0196, 13494.9764))
            it.addPoint(Point2d(671196.3657, 1230275.0454))
        }
        
        // Expand when setting the view bound so we can see the whole UK and the edge of the tile set
        if (displayVersion) {
            val extra = 0.5
            val extraX = extra * (bound.ur.x - bound.ll.x)
            val extraY = extra * (bound.ur.y - bound.ll.y)
            bound.addPoint(Point2d(bound.ll.x - extraX, bound.ll.y - extraY))
            bound.addPoint(Point2d(bound.ur.x + extraX, bound.ur.y + extraY))
        }

        return Proj4CoordSystem(projStr).apply {
            bounds = bound
        }
    }
    
    private fun makeTestLoader(activity: Activity, viewC: BaseController): QuadImageLoader {
        val bngCoordSystem = makeBNGCoordSystem(activity, false)
        val params = SamplingParams().apply {
            coordSystem = bngCoordSystem
            coverPoles = false
            edgeMatching = false
            minZoom = 0
            maxZoom = 10
            singleLevel = true
        }
        val info: TileInfoNew = TestTileInfo(0, 20)
        
        return QuadImageLoader(params, info, viewC).apply {
            setBaseDrawPriority(2000)
            tileFetcher = TestTileFetcher(viewC, "test").apply {
                alpha = 96
                start() // todo: does the loader clean this up?
            }
            debugMode = false
            // Additional tile boundaries for troubleshooting
            //setLoaderInterpreter(OvlDebugImageLoaderInterpreter().apply {
            //    setParentInterpreter(ImageLoaderInterpreter())
            //})
        }
    }
    
    private var baseCase: MaplyTestCase? = null
    private var loader: QuadImageLoader? = null
}